src/server/debug/server/server.go

```diff
@@ -601,7 +623,96 @@ func (s *debugServer) collectPipelineDumpFunc(pachClient *client.APIClient, limi
 		if err := s.collectCommits(tw, pachClient, client.NewRepo(pipelineInfo.Pipeline.Name), limit, prefix...); err != nil {
 			return err
 		}
-		return s.collectJobs(tw, pachClient, pipelineInfo.Pipeline.Name, limit, prefix...)
+		if err := s.collectJobs(tw, pachClient, pipelineInfo.Pipeline.Name, limit, prefix...); err != nil {
+			return err
+		}
+		if err := s.forEachWorkerLoki(pipelineInfo, func(pod string) error {
+			workerPrefix := join(podPrefix, pod)
+			if len(prefix) > 0 {
+				workerPrefix = join(prefix[0], workerPrefix)
+			}
+			userPrefix := join(workerPrefix, client.PPSWorkerUserContainerName)
+			if err := s.collectLogsLoki(tw, pod, client.PPSWorkerUserContainerName, userPrefix); err != nil {
+				return err
+			}
+			sidecarPrefix := join(workerPrefix, client.PPSWorkerSidecarContainerName)
+			return s.collectLogsLoki(tw, pod, client.PPSWorkerSidecarContainerName, sidecarPrefix)
+		}); err != nil {
+			name := "loki"
+			if len(prefix) > 0 {
+				name = join(prefix[0], name)
+			}
+			return writeErrorFile(tw, err, name)
+		}
+		return nil
+	}
+}
+
+func (s *debugServer) forEachWorkerLoki(pipelineInfo *pps.PipelineInfo, cb func(string) error) error {
+	pods, err := s.getWorkerPodsLoki(pipelineInfo)
+	if err != nil {
+		return err
+	}
+	for pod := range pods {
+		if err := cb(pod); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *debugServer) getWorkerPodsLoki(pipelineInfo *pps.PipelineInfo) (map[string]struct{}, error) {
+	queryStr := `{pipelineName="` + pipelineInfo.Pipeline.Name + `"}`
+	pods := make(map[string]struct{})
+	if err := s.queryLoki(queryStr, func(labels loki.LabelSet, _ string) error {
+		pod, ok := labels["pod"]
+		if !ok {
+			return errors.Errorf("pod label missing from loki label set")
+		}
+		pods[pod] = struct{}{}
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+	return pods, nil
+}
+
+// TODO: Need to bump up this default on the server side, it is too low as is.
+const maxLogs = 5000
+
+func (s *debugServer) queryLoki(queryStr string, cb func(loki.LabelSet, string) error) error {
+	c, err := s.env.GetLokiClient()
+	if err != nil {
+		return err
+	}
+	start := time.Time{}
+	end := time.Now()
+	for {
+		// TODO: Need a real context.
+		resp, err := c.QueryRange(s.env.Context(), queryStr, maxLogs, start, end, "FORWARD", 0, 0, true)
+		if err != nil {
+			return err
+		}
+		streams, ok := resp.Data.Result.(loki.Streams)
+		if !ok {
+			return errors.Errorf("resp.Data.Result must be of type loki.Streams")
+		}
+		var numLogs int
+		for _, stream := range streams {
+			for _, entry := range stream.Entries {
+				numLogs++
+				if entry.Timestamp.After(start) {
+					start = entry.Timestamp
+				}
+				if err := cb(stream.Labels, entry.Line); err != nil {
+					return err
+				}
+			}
+		}
+		if numLogs < maxLogs {
+			return nil
+		}
+		start = start.Add(time.Nanosecond)
```

Comment:
is the only difference between this and `lokiutil.QueryRange` the lack of sorting?

Reply:
Also the label sets in the callback.